1) Создал новую ветку add-friends-likes.

ViktorSalk opened the pull request:
2) Создал интерфейсы FilmStorage и UserStorage, в которых будут определены методы добавления, удаления и модификации объектов. 3) Создал классы InMemoryFilmStorage и InMemoryUserStorage, имплементирующие новые интерфейсы, и перенес всю логику хранения, обновления и поиска объектов. 4) Создал UserService и FilmService для обеспечения возможности пользователям добавлять друг друга в друзья и ставить фильмам лайки. 5) Создал класс ValidationService для фокусировки на проверке конкретных правил валидации данных в изоляции: Проверки длины описания фильма, Валидации имени фильма, Проверки email пользователя, Валидации логина и т.д. 6) Создал тесты на новую фичу: тесты фильма, тесты пользователя, тесты лайков, тесты друзей.

Viktor3642 committed:
1) Создал новую ветку add-friends-likes.
2) Создал интерфейсы FilmStorage и UserStorage, в которых будут определены методы добавления, удаления и модификации объектов.
3) Создал классы InMemoryFilmStorage и InMemoryUserStorage, имплементирующие новые интерфейсы, и перенес всю логику хранения, обновления и поиска объектов.
4) Создал UserService и FilmService для обеспечения возможности пользователям добавлять друг друга в друзья и ставить фильмам лайки.
5) Создал класс ValidationService для фокусировки на проверке конкретных правил валидации данных в изоляции: Проверки длины описания фильма, Валидации имени фильма, Проверки email пользователя, Валидации логина и т.д.
6) Создал тесты на новую фичу: тесты фильма, тесты пользователя, тесты лайков, тесты друзей.

Viktor3642 committed:
1) Внёс изменения для прохождения Postman теста.
2) Добавил проверку на нулевое значение лайков
3) Упростил добавление и удаление в filmStorage и InMemoryUserStorage

Viktor3642 committed:
1) Внёс изменения для прохождения Postman теста.
2) Обновил проверку на нулевое значение лайков
3) Обновил добавление и удаление в FilmService и InMemoryUserStorage

Viktor3642 committed:
1) Внёс изменения для прохождения Postman теста.
2) Обновил проверку на нулевое значение лайков
3) Обновил добавление и удаление в FilmService и UserController

Viktor3642 committed:
1) Внёс изменения для прохождения Postman теста.
2) Обновил проверку на нулевое значение лайков
3) Обновил добавление и удаление в User, UserService и ValidationServiceTests

Viktor3642 committed:
1) Внёс изменения для прохождения Postman теста.
2) Добавил ErrorHandler
3) Обновил  UserController и FilmController

Viktor3642 committed:
1) Внёс изменения для прохождения Postman теста.
2) Добавил ErrorHandler
3) Обновил  UserController и FilmController

Viktor3642 committed:
1) Внёс изменения для прохождения Postman теста.
2) Добавил ErrorHandler
3) Обновил  UserController и FilmController

Viktor3642 committed:
1) Внёс изменения для прохождения Postman теста.
2) Изменил UserController для возвращения кода 404 при попытке обновить несуществующего пользователя

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,78 +1,79 @@
         }
-        if (films.containsKey(film.getId())) {
-            Film oldFilm = films.get(film.getId());
+        if (!filmService.getAllMapFilms().containsKey(film.getId())) {
```
Код со строк 47-50 является бизнес логикой приложения, его нужно вынести на уровень сервиса

ViktorSalk replied:
Вынес на уровень сервиса бизнес логику из FilmController.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,78 +1,79 @@
     @PutMapping
-    @ResponseStatus(HttpStatus.OK)
     public Film updateFilm(@Valid @RequestBody Film film) {
         if (film.getId() == null) {
```
Лучше добавить аннотацию валидации `@NotNull` над полем `id` в классе `Film`

ViktorSalk replied:
Убрал лишний код (при добавлении аннотации валидации @NotNull в класс Film срабатывают ошибки в прохождении POSTMAN теста.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,95 +1,73 @@
-        if (newUser.getId() == null) {
-            log.error("Не введен id пользователя при изменении");
+    public User updateUser(@Valid @RequestBody User user) {
+        if (user.getId() == null) {
```
Нужно добавить аннотацию `@NotNull` над полем `id` класса User

ViktorSalk replied:
Убрал лишний код (при добавлении аннотации валидации @NotNull в класс User срабатывают ошибки в прохождении POSTMAN теста.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,95 +1,73 @@
+        if (user.getId() == null) {
             throw new ValidationException("id пользователя должен быть указан");
         }
+        if (!userService.getAllUserMap().containsKey(user.getId())) {
```
Код со строк 40-43 нужно перенести в сервис

ViktorSalk replied:
Вынес на уровень сервиса бизнес логику из UserController.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,95 +1,73 @@
-        throw new UserNotFoundException("Пользователь с id =" + newUser.getId() + " не найден");
+    @ExceptionHandler(UserNotFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public Map<String, String> handleUserNotFound(final UserNotFoundException e) {
```
Есть класс `GlobalExceptionHandler`, где отлавливаются все ошибки приложения. Данный код нужно удалить

ViktorSalk replied:
Расширил класс GlobalExceptionHandler, в котором отлавливаются ошибки приложения, убрал дубль из UserController.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,78 +1,79 @@
+
+    @ExceptionHandler(UserNotFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public Map<String, String> handleUserNotFound(final UserNotFoundException e) {
```
Есть класс `GlobalExceptionHandler`, где отлавливаются все ошибки приложения. Данный код нужно удалить

ViktorSalk replied:
Убрал дубль из FilmController.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -24,4 +26,12 @@ public class Film {
     @NotNull(message = "Длительность не может быть пустой")
     @PositiveOrZero(message = "Длительность фильма должна быть больше или равна нулю")
     private Integer duration;
+    private Set<Long> idUserLikes = new HashSet<>();
```
Отличный выбор коллекции 👍

ViktorSalk replied:
Спасибо))

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -33,4 +37,12 @@ public void setLogin(String login) {
+        return friends;
+    }
+
+    public void setFriends(Set<Long> friends) {
```
Геттер с 41 строки и сеттер с 45 автоматически сгенерируются с помощью lombok, так как над классом строит аннотация `@Data`. Данный код нужно удалить

ViktorSalk replied:
Удалил лишние getFriends и setFriends из User.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/ErrorHandler.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.Map;
+
+@RestControllerAdvice
+public class ErrorHandler {
```
У тебя есть класс `GlobalExceptionHandler`, который занимается отлавливанием ошибок по всему приложению. Данный класс нужно удалить, а `GlobalExceptionHandler` нужно расширить

ViktorSalk replied:
Удалил не используемый класс ErrorHandler.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,86 @@
+import ru.yandex.practicum.filmorate.storage.FilmStorage;
+import ru.yandex.practicum.filmorate.storage.UserStorage;
+
+import java.util.*;
```
_Саморазвитие_
В коммерческой разработке не принято использовать `wildcard` в импортах.
Вот отличная [статья](https://struchkov.dev/blog/ru/using-wildcard-imports-in-java/#:~:text=%D0%98%D1%81%D0%BF%D0%BE%D0%BB%D1%8C%D0%B7%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5%20%D0%BF%D0%BE%D0%B4%D1%81%D1%82%D0%B0%D0%BD%D0%BE%D0%B2%D0%BE%D1%87%D0%BD%D1%8B%D1%85%20%D0%B8%D0%BC%D0%BF%D0%BE%D1%80%D1%82%D0%BE%D0%B2%20%D0%BD%D0%B8%D0%BA%D0%B0%D0%BA%20%D0%BD%D0%B5,%D1%81%D1%82%D0%BE%D0%BB%D0%BA%D0%BD%D1%83%D1%82%D1%8C%D1%81%D1%8F%20%D1%81%20%D0%BF%D1%80%D0%BE%D0%B1%D0%BB%D0%B5%D0%BC%D0%BE%D0%B9%20%D0%BA%D0%BE%D0%BD%D1%84%D0%BB%D0%B8%D0%BA%D1%82%D0%B0%20%D0%B8%D0%BC%D0%B5%D0%BD.) почему и как настроить идею, чтобы избавиться от `*`
Кстати, в IDEA есть отличное сочетание клавиш `Ctrl + Alt + O` - при нажатии сочетания в нужном классе автоматически оптимизируются импорты и удаляются ненужные

ViktorSalk replied:
Спасибо))
Убрал wildcard из импортов проекта.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,86 @@
+    private final FilmStorage filmStorage;
+    private final UserStorage userStorage;
+
+    private final LikeComparator likeComparator = new LikeComparator();
```
Нигде не используется

ViktorSalk replied:
Удалил не используемый LikeComparator из FilmService.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    public void addLike(Long userId, Long id) {
+        Film film = filmStorage.getFilm(id);
```
Код со строк 38-48 повторяется на строках 57-67. Нужно его выделить в отдельный метод и переиспользовать

ViktorSalk replied:
Выделил в отдельный метод и переиспользовал повторяющиеся строки кода в FilmService.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/LikeComparator.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.util.Comparator;
+
+public class LikeComparator implements Comparator<Film> {
```
Данный компаратор нигде не используется. Его нужно либо удалить, либо использовать

ViktorSalk replied:
Удалил не используемый LikeComparator.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,86 @@
+
+    public Collection<Film> getPopularFilms(int count) {
+        List<Film> films = new ArrayList<>(filmStorage.getAllFilms());
+        films.sort((f1, f2) -> Integer.compare(
```
Нужно избавиться от однобуквенных переменных и чисел в названии переменных. Из-за них код становится сложнее читать; лучше использовать `film` и `filmToCompare`

ViktorSalk replied:
Использовал film и filmToCompare вместо однобуквенных переменных в FilmService.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        List<User> friends = new ArrayList<>();
+        for (Long friendId : user.getFriends()) {
```
Нужно переписать цикл с помощью [stream api](https://www.baeldung.com/java-8-streams)

ViktorSalk replied:
Переписал циклы в стиле Stream API в UserService.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,114 @@
+        commonFriendIds.retainAll(otherUser.getFriends());
+
+        List<User> commonFriends = new ArrayList<>();
+        for (Long friendId : commonFriendIds) {
```
Нужно переписать цикл с помощью [stream api](https://www.baeldung.com/java-8-streams)

ViktorSalk replied:
Переписал циклы в стиле Stream API в UserService.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/ValidationService.java:
```diff
@@ -0,0 +1,138 @@
+import java.util.Set;
+
+@Slf4j
+public class ValidationService {
```
Этот класс нужно удалить. А также нужно удалить все тексты, которые тестируют данный класс. Тесты должны тестировать реальную валидацию.
Для тестирования спринговой валидации можно воспользоваться [валидатором](https://stackoverflow.com/questions/29069956/how-to-test-validation-annotations-of-a-class-using-junit)

ViktorSalk replied:
Удалил не используемый класс ValidationService и заменил тесты.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmStorage.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.Map;
+
+public interface FilmStorage {
+    Collection<Film> getAllFilms();
```
Нужно возвращать не `Collection` менее абстрактный интерфейс - `List`, так будет проще перейти на использование реальной базы данных в следующем спринте

ViktorSalk replied:
Использовал List вместо Collection в FilmStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public void setAllFilmsMap(Map<Long, Film> allFilmsMap) {
```
Данный метод нужно удалить. А на уровнь класса добавить аннотацию `@RequiredArgsConstructor`, которая сгенерирует конструктор для всех `final` полей класса

ViktorSalk replied:
Удалил лишний метод, добавив аннотацию генерирующую конструктор для final полей класса в InMemoryFilmStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,65 @@
+@Component
+public class InMemoryFilmStorage implements FilmStorage {
+
+    private Map<Long, Film> allFilms = new HashMap<>();
```
Поле должно быть `final`

ViktorSalk replied:
Сделал класс InMemoryFilmStorage final.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public Map<Long, Film> getAllFilmsMap() {
```
Этот метод нужно удалить и если нужно проверить, что фильм существует в хранилище, то просто использовать `getFilm` вместо, так как он выбрасывает ошибку, если фильма с данным `id` нет в мапе `allFilms`

ViktorSalk replied:
Удалил лишние методы из InMemoryFilmStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserStorage.java:
```diff
@@ -0,0 +1,30 @@
+
+
+public interface UserStorage {
+    Collection<User> getAllUsers();
```
Нужно возвращать `List`

ViktorSalk replied:
Использовал List вместо Collection в UserStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,84 @@
+
+    private Map<Long, User> allUsers = new HashMap<>();
+
+    private Map<Long, Set<User>> friendsMap = new HashMap<>();
```
Хранить мапу друзей дополнительно не нужно. Все друзья каждого из пользователей хранятся в классе `User` в поле `friends` - нужно использовать его

ViktorSalk replied:
Удалил лишнюю мапу из InMemoryUserStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,84 @@
+    private Map<Long, Set<User>> friendsMap = new HashMap<>();
+
+    @Override
+    public void setFriendsMap(Map<Long, Set<User>> friendsMap) {
```
Нужно удалить данный метод и использовать `@RequiredArgsConsructor`

ViktorSalk replied:
Удалил лишний метод, добавив аннотацию генерирующую конструктор для final полей класса в InMemoryUserStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,84 @@
+@Component
+public class InMemoryUserStorage implements UserStorage {
+
+    private Map<Long, User> allUsers = new HashMap<>();
```
Поле должно быть `final`

ViktorSalk replied:
Сделал класс InMemoryUserStorage final.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public Map<Long, User> getCollectionAllUsers() {
```
Этот метод нужно удалить

ViktorSalk replied:
Удалил лишние методы из InMemoryUserStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public Map<Long, Set<User>> getFriendsMap() {
```
Этот метод нужно удалить

ViktorSalk replied:
Удалил лишние методы из InMemoryUserStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public void setCollectionAllUsers(Map<Long, User> allUsers1) {
```
Этот метод нужно удалить

ViktorSalk replied:
Удалил лишние методы из InMemoryUserStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public Set<User> getUserFriends(long id) {
```
Этот метод нужно удалить

ViktorSalk replied:
Удалил лишние методы из InMemoryUserStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public void updateUsersFriends(long id, Set<User> userSetFriends) {
```
Этот метод нужно удалить

ViktorSalk replied:
Удалил лишние методы из InMemoryUserStorage.

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/ValidationServiceTests.java:
```diff
@@ -0,0 +1,207 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+@SpringBootTest
+public class ValidationServiceTests {
```
Данные тесты нужно удалить, так как они тестируют не работающую валидацию. Взамен нужно написать тесты, которые будет тестировать реальную валидацию. Как это сделать написала в комментарии к `ValidationService`

ViktorSalk replied:
Удалил не используемый класс ValidationService и заменил тесты на ValidationTests с проверкой валидации и ServiceTests с проверкой на работу с друзьями, с лайками, популярными фильмами.

Viktor3642 committed:
1) Вынес на уровень сервиса бизнес логику из FilmController.
2) Убрал лишний код (при добавлении аннотации валидации @NotNull в класс Film срабатывают ошибки в прохождении POSTMAN теста.
3) Убрал лишний код (при добавлении аннотации валидации @NotNull в класс User срабатывают ошибки в прохождении POSTMAN теста.
4) Вынес на уровень сервиса бизнес логику из UserController.
5) Расширил класс GlobalExceptionHandler, в котором отлавливаются ошибки приложения, убрал дубль из UserController.
6) Удалил лишние методы из InMemoryUserStorage.
7) Удалил не используемый класс LikeComparator.
8) Удалил лишние getFriends и setFriends из User.
9) Убрал wildcard из импортов проекта.
10) Удалил не используемый LikeComparator из FilmService.
11) Выделил в отдельный метод и переиспользовал повторяющиеся строки кода в FilmService.
12) Использовал film и filmToCompare вместо однобуквенных переменных в FilmService.
13) Переписал циклы в стиле Stream API в UserService.
14) Удалил не используемый класс ValidationService.
15) Удалил не используемый класс ValidationServicetest.
16) Создал класс Validationtest для тестирования спринговой валидации.
17) Использовал List вместо Collection в FilmStorage.
18) Сделал класс InMemoryFilmStorage final.
19) Удалил лишний метод, добавив аннотацию генерирующую конструктор для final полей класса в InMemoryFilmStorage.
20) Удалил лишние методы из InMemoryFilmStorage.
21) Сделал класс InMemoryUserStorage final.
22) Удалил лишнюю мапу из InMemoryUserStorage.
23) Удалил лишний метод, добавив аннотацию генерирующую конструктор для final полей класса в InMemoryUserStorage.
24) Удалил лишние методы из InMemoryUserStorage.
25) Использовал List вместо Collection в UserStorage.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,78 +1,68 @@
     @GetMapping
     public List<Film> findAll() {
-        return new ArrayList<>(films.values());
+        return new ArrayList<>(filmService.getAllFilms());
```
Чтобы не кастовать Collection в List, на уровне сервиса нужно возращать интерфейс List вместо [Collection](https://github.com/ViktorSalk/java-filmorate/blob/a10e8115a48e47be151fa62eca18d9fb5615513d/src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java#L24)

ViktorSalk replied:
Заменил Collection на List в FilmService.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,78 +1,68 @@
-        log.error("Фильм с id = {} не найден", film.getId());
-        throw new UserNotFoundException("Фильм с id = " + film.getId() + " не найден");
+    @GetMapping("/popular")
+    public Collection<Film> getPopularFilms(@RequestParam(required = false, defaultValue = "10") Integer count) {
```
Давай здесь тоже будем возвращать пользователю `List`

ViktorSalk replied:
Заменил Collection на List в FilmService и FilmController.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,95 +1,67 @@
     @GetMapping
     public List<User> findAll() {
-        return new ArrayList<>(users.values());
+        return new ArrayList<>(userService.getAllUsers());
```
Нужно на уровне сервиса возвращать List вместо [Collection](https://github.com/ViktorSalk/java-filmorate/blob/a10e8115a48e47be151fa62eca18d9fb5615513d/src/main/java/ru/yandex/practicum/filmorate/service/UserService.java#L24)

ViktorSalk replied:
Заменил Collection на List в UserService.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,95 +1,67 @@
-        throw new UserNotFoundException("Пользователь с id =" + newUser.getId() + " не найден");
+    @GetMapping("/{id}/friends/common/{otherId}")
+    public List<User> getCommonFriends(@PathVariable Long id, @PathVariable Long otherId) {
+        return new ArrayList<>(userService.generalFriends(id, otherId));
```
Замени [Collection](https://github.com/ViktorSalk/java-filmorate/blob/a10e8115a48e47be151fa62eca18d9fb5615513d/src/main/java/ru/yandex/practicum/filmorate/service/UserService.java#L92) на List

ViktorSalk replied:
Заменил Collection на List в UserService

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/GlobalExceptionHandler.java:
```diff
@@ -8,6 +8,12 @@
 @RestControllerAdvice
 public class GlobalExceptionHandler {
 
+    public static class FilmNotFoundException extends RuntimeException {
```
Эта ошибка нигде не выбрасывается, чтобы ее отлавливать в `ExceptionHandler`. Данный код и код со строк 29-34 - бесполезный

ViktorSalk replied:
Удалил неиспользуемый код в GlobalExceptionHandler.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private long getNextId() {
+        long currentMaxId = allUsers.keySet().stream().mapToLong(id -> id).max().orElse(0);
```
Стримы принято бить по строкам:
```suggestion
        long currentMaxId = allUsers.keySet().stream()
            .mapToLong(id -> id)
            .max()
            .orElse(0);
```

ViktorSalk replied:
Сделал стрим более читаемым.

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/service/ServiceTests.java:
```diff
@@ -0,0 +1,142 @@
+    class FilmLikesTests {
+        @Test // Тесты для работы с лайками
+        void testLikeOperations() {
+            Film film = new Film();
```
Вместо того, чтобы в каждом тесте создавать одинаковые объекты, их можно вынести в блок `@BeforeEach` и избавиться от дублирований кода

ViktorSalk replied:
Удалил дубли, вынеся их блоком @BeforeEach в ServiceTests.

Viktor3642 committed:
1) Заменил Collection на List в FilmService.
2) Заменил Collection на List в UserService.
3) Заменил возвращение пользователю getPopularFilms интерфейс List в FilmService.
4) Удалил неиспользуемый код в GlobalExceptionHandler.
5) Сделал стрим более читаемым.
6) Удалил дубли, вынеся их блоком @BeforeEach в ServiceTests.java
